lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,178 @@
+    property:unset()
+    return result
+end
+
+
+--- Get GTK+3 theme variables from GtkStyleContext
+-- @treturn table subj
+function gtk.get_theme_variables()
+  if gtk.cached_theme_variables then
+    return gtk.cached_theme_variables
+  end
+
+  local result = {}
+  local lgi = require('lgi')
+  local Gtk = lgi.Gtk
+  local window
+  pcall(function()
+    window = Gtk.Window{
+      --on_destroy = Gtk.main_quit,
+    }
+  end)
+  if not window then
+    gears_debug.print_warning("Seems like GTK+3 is not installed or theme is not set correctly.")
+    return nil
+  end
+  --window:set_override_redirect(true)
+  local style_context = window:get_style_context()
+
+  local font = style_context:get_font(Gtk.StateFlags.NORMAL)
+  --result.font_family = font:get_family()
+  --result.font_size = font:get_size()/1024
+  result.font = font:to_string()
+  for result_key, style_context_property in pairs({
+    bg_color="background-color",
+    fg_color="color",
+  }) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        style_context, style_context_property, Gtk.StateFlags.NORMAL
+    )
+  end
+
+  local entry = Gtk.Entry()
+  local entry_style_context = entry:get_style_context()
+  for result_key, style_context_property in pairs({
+    border_radius="border-radius",
+    border_width="border-top-width",
+  }) do
+    local property = entry_style_context:get_property(style_context_property, Gtk.StateFlags.NORMAL)
+    result[result_key] = property.value
+    property:unset()
+  end
+  for result_key, style_context_property in pairs({
+    base_color="background-color",
+    text_color="color",
+  }) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        entry_style_context, style_context_property, Gtk.StateFlags.NORMAL
+    )
+  end
+  entry:destroy()
+
+  local toggle_button = Gtk.ToggleButton()
+  toggle_button:set_active()
+  local toggle_button_style_context = toggle_button:get_style_context()
+  for result_key, style_context_property in pairs({
+    selected_bg_color="background-color",
+    selected_fg_color="color",
+  }) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        toggle_button_style_context, style_context_property, Gtk.StateFlags.ACTIVE
+    )
+  end
+  toggle_button:destroy()
+
+  local button = Gtk.Button()
+  local button_style_context = button:get_style_context()
+  for result_key, style_context_property in pairs({
+    border_radius="border-radius",
+    border_width="border-top-width",
+  }) do
+    local property = button_style_context:get_property(style_context_property, Gtk.StateFlags.NORMAL)
+    result[result_key] = property.value
+    property:unset()
+  end
+  for result_key, style_context_property in pairs({
+    button_bg_color="background-color",
+    button_fg_color="color",
+  }) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        button_style_context, style_context_property, Gtk.StateFlags.NORMAL
+    )
+  end
+  button:destroy()
```

Comment:
Can the above code be made more readable by e.g. moving it into a table?
```lua
{
object = Gtk.Button(),
properties = { 
  Gtk.StateFlags.NORMAL = {
    border_radius = "border-radiues",
    etc
  }
}
```
That way, all that code is only needed once instead of copy&pasted three times.